### PR TITLE
[WIP] CHI-566 edit case & dedup spike

### PR DESCRIPTION
**Pull request by stephenhand**
** DO NOT REVIEW!! ** 

The work here is now in a 'refactor only' PR here: https://github.com/techmatters/flex-plugins/pull/613 and an upcoming PR that will implement the edit functionality. I'm only keeping this PR open so I can continue any ongoing discussions already started with @GPaoloni 

PLEASE READ https://bugs.benetech.org/browse/CHI-566?focusedCommentId=239639&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-239639 before reviewing! It will give a bit more context on what is already known needs to be done / could be an issue - could save you some time :-)

## Description

* Adds 'edit' functionality to individual elements of a case
* Deduplicates the case form section components (Household.tsx, Perpetrator.tsx etc.) and the Add / Edit forms (AddHousehold.tsx, AddPerpetrator.tsx etc)
* Adds UUID identifiers to all case components

### Checklist
- [ X ] Corresponding issue has been opened
- [ X ] New tests added
- [ ] Strings are localized

### Related Issues
Fixes CHI-566

### Verification steps

* Try editing each type of item in an existing case
* Try editing each type of item in a new case
* Regression test adding each type of item in an existing case
* Regression test adding each type of item in a new case
* Regression test editing top level info in an existing case
* Regression test adding top level info in a new case
* Regression test viewing each type of item in an existing case
* Regression test viewing each type of item in a new case
* Test new & existing permissions around cases

### CODE REVIEW

**Review by GPaoloni — COMMENTED** [review]
This looks like an amazing improvement on the re-usability of the case components you are abstracting. To be honest the PR requires a lot of cognitive effort to understand everything that's going on so maybe a more progressive split of the refactor may be helpful, but I'm not even sure if that's possible. Also it looks to me that the routing types and the routes reducer that we've been using is causing troubles. It may be a good time to step back and think if this is really the best solution or if we should rethink the way we handle our screens. I need to look again at this again, I'm just dropping it for now cause it's been more than an hour of reading it and I'm losing the focus haha.

Also one question: what's the reason for the uuid addition to the case items? The ability to index them so we can determine what particular item are editing?

File: plugin-hrm-form/src/components/case/AddEditCaseItem.tsx [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  applyTemporaryInfoToCase,
+  customFormHandlers,
+  reactHookFormOptions,
+  // eslint-disable-next-line sonarjs/cognitive-complexity
+}) => {
+  const firstElementRef = useFocus();
+
+  const { temporaryCaseInfo } = connectedCaseState;
+
+  const isEditingExistingItem = (): boolean => isEditTemporaryCaseInfo(temporaryCaseInfo);
```
[the remark this answers is not in the document]
Comment (stephenhand) [reply plugin-hrm-form/src/components/case/AddEditCaseItem.tsx:100]: I just think the function is more readable than calling `isEditTemporaryCaseInfo(temporaryCaseInfo)` everywhere.

Just FYI, modern JavaScript engines will almost certainly not be redefining this function on every call. Since local functions are a very common pattern in JS, engines like V8 and SpiderMonkey will cache the implementation of the local function after the first call and reuse it from then on. In fact, a trivial function like this would most likely be 'inlined', i.e. the call to the function replaced with a copy of the implementation wherever it is called, like ye olde `inline` keyword used to do in C (modern C compilers pretty much ignore it).

Basically, inner functions in functions that are called a lot are not a big performance consideration, organise function hierarchies for readability & maintainability, the JS engine will figure out the optimisation in most cases.

File: plugin-hrm-form/src/components/case/AddEditCaseItem.tsx [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  const save = async shouldStayInForm => {
+    const { info, id } = connectedCaseState.connectedCase;
+    const form = transformValues(formDefinition)(getTemporaryFormContent(temporaryCaseInfo));
+    const createdAt = new Date().toISOString();
+    const { workerSid } = getConfig();
+    let newInfo: CaseInfo;
+    if (isEditTemporaryCaseInfo(temporaryCaseInfo)) {
+      newInfo = applyTemporaryInfoToCase(
+        info,
+        { ...temporaryCaseInfo.info, form, id: temporaryCaseInfo.info.id ?? uuidV4() },
```
[the remark this answers is not in the document]
Comment (stephenhand) [reply plugin-hrm-form/src/components/case/AddEditCaseItem.tsx:158]: The UUID is not used right now, all the logic is based off the array position. It's just added to the models in this PR because we will likely use it in the future - especially if we move case sections out of the JSONB field in the DB and into their own table, they will have their own unique identifiers then. 

When we do that we will have to figure out the specifc upgrade sequence, but I think it's good for the front end model to at least be aware of the notion of a unique identifier. It also brings all the other case item models in line with Documents, which already had a unique identifier

File: plugin-hrm-form/src/components/case/AddEditCaseItem.tsx [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    const updatedCase = await updateCase(id, { info: newInfo });
+    setConnectedCase(updatedCase, task.taskSid, true);
+    if (shouldStayInForm && isAddTemporaryCaseInfo(temporaryCaseInfo)) {
+      const blankForm = formDefinition.reduce(createStateItem, {});
+      methods.reset(blankForm); // Resets the form.
+      updateTempInfo({ screen: temporaryCaseInfo.screen, info: {} }, task.taskSid);
+      changeRoute({ route, subroute: temporaryCaseInfo.screen }, task.taskSid);
+    }
```
[the remark this answers is not in the document]
Comment (stephenhand) [reply plugin-hrm-form/src/components/case/AddEditCaseItem.tsx:173]: Agreed, but in the interests of keeping the PR size down (!) I left it where it was for now

File: plugin-hrm-form/src/components/case/Case.tsx [hunk 4]
```diff
@@ -259,44 +250,20 @@ const Case: React.FC<Props> = props => {
...
-  };
+  type CaseItemInfo<T extends TemporaryCaseInfo> = T['info']; // A bit redundant but looks cleaner than the anonymous subtype reference syntax
 
-  const onClickViewPerpetrator = (perpetrator: PerpetratorEntry) => {
-    props.updateTempInfo({ screen: 'view-perpetrator', info: perpetrator }, props.task.taskSid);
-    props.changeRoute({ route, subroute: 'view-perpetrator' }, props.task.taskSid);
+  const onCaseItemActionClick = <T extends TemporaryCaseInfo>(
+    targetSubroute: CaseItemRoute<T>,
+  ): ((entry: CaseItemInfo<T>) => void) => (entry: CaseItemInfo<T>) => {
+    props.updateTempInfo({ screen: targetSubroute, info: entry } as TemporaryCaseInfo, props.task.taskSid);
```
[the remark this answers is not in the document]
Comment (stephenhand) [reply plugin-hrm-form/src/components/case/Case.tsx:260]: You're right, casts should be avoided where possible & I tried here but I didn't see a good option without even more big code changes being piled into the PR. 

You can have a look, but honestly, I wouldn't devote too much effort to it. My feeling is that this particular part of the type model is doing more harm than good - it gets in the way of refactoring and offers no valuable extra type checking.

I think if we were going to do more refactoring here (and I think we should) I would prefer to devote that energy to doing away with the practice of passing blobs of state from one case component to another completely. I think it would be better to just express the specific part of the case we want to edit as part of the route (e.g. `case/4321/perpetrators/3/?edit` or `case/1234/households?add` or whatever the convention is) and have the component use the information in the route to read the data it needs from the state independently. 

At that point routes can just be strings and we don't need to try and work with 'route and associated state blob type' pairings. Also the components are much more standalone and easier to invoke from any part of the application. There may of course, be good reasons why the above plan won't work - I'm still grokking this code :-)

**Comment by stephenhand** [comment]
> This looks like an amazing improvement on the re-usability of the case components you are abstracting. To be honest the PR requires a lot of cognitive effort to understand everything that's going on so maybe a more progressive split of the refactor may be helpful, but I'm not even sure if that's possible. Also it looks to me that the routing types and the routes reducer that we've been using is causing troubles. It may be a good time to step back and think if this is really the best solution or if we should rethink the way we handle our screens. I need to look again at this again, I'm just dropping it for now cause it's been more than an hour of reading it and I'm losing the focus haha.
> 
> Also one question: what's the reason for the uuid addition to the case items? The ability to index them so we can determine what particular item are editing?

I started by trying to only do a subset of the refactor you see here, but it was hard to draw a neat line between 1 part of the refactor and the next - you could do it but it would require a bunch of temporary fix ups to get it to still run, which would be removed in the next part, so it made more sense to do it all together.

One thing I can do is split the refactor and the edit functionality (which needs reworking slightly to match Jana's mockups anyway). Hopefully the refactor PR by itself will be a little easier to review, if not I can look at trying to split it up further

File: plugin-hrm-form/src/___tests__/components/case/ViewCaseItem.test.tsx [hunk 5]
```diff
@@ -1,3 +1,4 @@
+// @ts-ignore
```
[the remark this answers is not in the document]
Comment (GPaoloni) [reply plugin-hrm-form/src/___tests__/components/case/ViewCaseItem.test.tsx:1]: Oh sorry I got confused because of it being at the very beginning of the file!

File: plugin-hrm-form/src/components/case/AddEditCaseItem.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (GPaoloni) [reply plugin-hrm-form/src/components/case/AddEditCaseItem.tsx:100]: You are right about the function definition not being a performance consideration at all. Also, I'm really interested on learning more on this topic, if you want, this sounds like an awesome all eng meeting proposal  to talk about :nerd_face:
